lib/client.js

```diff
@@ -934,23 +1081,49 @@ MqttClient.prototype._handlePublish = function (packet, done) {
   var that = this
 
   switch (qos) {
-    case 2:
-      this.incomingStore.put(packet, function () {
-        that._sendPacket({cmd: 'pubrec', messageId: mid}, done)
-      })
+    case 2: {
+      if (this.options.customHandleAcks) {
+        this.options.customHandleAcks(topic, message, packet, function (code) {
+          if (code) {
+            that._sendPacket({cmd: 'pubrec', messageId: mid, reasonCode: code}, done)
+          } else {
+            that.incomingStore.put(packet, function () {
+              that._sendPacket({cmd: 'pubrec', messageId: mid}, done)
+            })
+          }
+        })
+      } else {
+        this.incomingStore.put(packet, function () {
+          that._sendPacket({cmd: 'pubrec', messageId: mid}, done)
+        })
+      }
       break
-    case 1:
+    }
+    case 1: {
       // emit the message event
-      this.emit('message', topic, message, packet)
-      this.handleMessage(packet, function (err) {
-        if (err) {
-          return done && done(err)
-        }
-        // send 'puback' if the above 'handleMessage' method executed
-        // successfully.
-        that._sendPacket({cmd: 'puback', messageId: mid}, done)
-      })
+      if (this.options.customHandleAcks) {
+        this.options.customHandleAcks(topic, message, packet, function (code) {
+          if (!code) { this.emit('message', topic, message, packet) }
+          that.handleMessage(packet, function (err) {
+            if (err) {
+              return done && done(err)
+            }
+            that._sendPacket({cmd: 'puback', messageId: mid, reasonCode: code}, done)
+          })
+        })
+      } else {
+        this.emit('message', topic, message, packet)
+        this.handleMessage(packet, function (err) {
+          if (err) {
+            return done && done(err)
+          }
+          // send 'puback' if the above 'handleMessage' method executed
+          // successfully.
+          that._sendPacket({cmd: 'puback', messageId: mid}, done)
+        })
+      }
       break
+    }
```

Comment:
All this function becomes very complicated with the `customHandleHack` function. Can you please refactor so that there is _always_ a default for that, and we avoid all those ifs?

Reply:
refactored, check pls